### PR TITLE
FOGL-1566 getCategory root param fixes

### PR DESCRIPTION
**Pull request by ashish-jabble**

### CODE REVIEW

File: python/foglamp/common/configuration_manager.py [hunk 1]
```diff
@@ -210,17 +210,44 @@ def get_entry_val(k):
         return category_info
 
     async def _read_all_groups(self, root):
-        payload = PayloadBuilder().SELECT("key", "description", "value").chain_payload()
-        if root is True:
-            payload = PayloadBuilder(payload).WHERE(["value", "=", "{}"]).payload()
+        group_info = []
+        if root:
+            # SELECT DISTINCT parent FROM category_children WHERE parent NOT IN (SELECT child FROM category_children);
```
Comment (MarkRiddoch) [inline python/foglamp/common/configuration_manager.py]: Please see latest comment in Jira, it is the list of categories that is the base list to remove entries from, as you may have a root category that has no children and hence is not a parent.

**Comment by amarendra-dianomic** [comment]
@ashish-jabble please examine below snippet for suitability:
```
    async def _read_all_groups(self, root):
        group_info = []

        # TODO: FOGL-668 Need support for subquery
        payload = PayloadBuilder().SELECT("parent").payload()
        result = await self._storage.query_tbl_with_payload('category_children', payload)
        parents = set([v['parent'] for v in result['rows']])

        payload2 = PayloadBuilder().SELECT("child").payload()
        result2 = await self._storage.query_tbl_with_payload('category_children', payload2)
        children = set([v['child'] for v in result2['rows']])

        if root:
            # SELECT DISTINCT parent FROM category_children WHERE parent NOT IN (SELECT child FROM category_children);
            diffs = parents - children  # returns the set of elements in parents, but not in children
        else:
            # SELECT DISTINCT child FROM category_children WHERE child IN (SELECT parent FROM category_children);
            diffs = parents & children  # returns the set of elements common to both in parents and children

        for item in diffs:
            p = PayloadBuilder().SELECT("key", "description").WHERE(["key", "=", item]).payload()
            r = await self._storage.query_tbl_with_payload('configuration', p)
            group_info.append((r['rows'][0]['key'], r['rows'][0]['description']))

        return group_info
```

File: python/foglamp/common/configuration_manager.py [hunk 2]
```diff
@@ -210,45 +210,26 @@ def get_entry_val(k):
...
-            payload2 = PayloadBuilder().SELECT("parent").payload()
-            result2 = await self._storage.query_tbl_with_payload('category_children', payload2)
-
-            def is_parent(child):
-                if child in [row['parent'] for row in result2['rows']]:
-                    return True
-                return False
+        # SELECT DISTINCT child FROM category_children
+        payload2 = PayloadBuilder().SELECT("child").DISTINCT(["child"]).payload()
+        result2 = await self._storage.query_tbl_with_payload('category_children', payload2)
```
Comment (praveen-garg) [inline python/foglamp/common/configuration_manager.py]: payload2 -> unique_category_children_sql_payload
result2 -> unique_category_children

File: python/foglamp/common/configuration_manager.py [hunk 3]
```diff
@@ -210,45 +210,26 @@ def get_entry_val(k):
...
-            payload2 = PayloadBuilder().SELECT("child").payload()
-            result2 = await self._storage.query_tbl_with_payload('category_children', payload2)
-
-            def has_children(parent):
-                if parent in [row['child'] for row in result2['rows']]:
-                    return True
-                return False
+        # SELECT key, description FROM configuration
+        payload = PayloadBuilder().SELECT("key", "description").payload()
+        result = await self._storage.query_tbl_with_payload('configuration', payload)
```
Comment (praveen-garg) [inline python/foglamp/common/configuration_manager.py]: all_categories

**Pull request merged**